File: pc-common/src/main/java/com/cloud/pc/utils/UrlProbe.java
/*
 * Copyright (c) 2025 Yangagile. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.cloud.pc.utils;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;

/*
 *
 */
public class UrlProbe {
    private static final Logger LOG = LoggerFactory.getLogger(UrlProbe.class);
    public long probe_min_period_ms = ComUtils.getProps("pc.probe.min.period.ms",
            5*60*1000L, Long::valueOf);

    private volatile List<UrlStats> urlList = new ArrayList<>();
    private IUrlProbeFunction probeFunction;
    private final AtomicBoolean isRunning;
    private final AtomicLong lastProbeTime;

    public interface IUrlProbeFunction {
        List<String> getUrlList(String url);
    }

    class UrlStats {
        UrlStats(String url, long responseTime, boolean active) {
            this.url = url;
            this.responseTime = responseTime;
            this.active = new AtomicBoolean(active);
        }
        String url;
        long responseTime;
        AtomicBoolean active;

        public long getResponseTime() {
            return responseTime;
        }
    }

    public UrlProbe(String baseUrl, IUrlProbeFunction probeFunction) {
        List<String> urls = probeFunction.getUrlList(baseUrl);
        for (String url : urls) {
            urlList.add(new UrlStats(url, 0, true));
        }
        this.probeFunction = probeFunction;
        this.isRunning = new AtomicBoolean(false);
        this.lastProbeTime = new AtomicLong(0);
    }

    public String getUrl() {
        List<UrlStats> tmpList = urlList;
        for (UrlStats urlStats : tmpList) {
            if (urlStats.active.get()) {
                return urlStats.url;
            }
        }
        // if all are inactive, return the first
        for (UrlStats urlStats : tmpList) {
            return urlStats.url;
        }
        probe(false);
        return null;
    }

    public void reportFail(String url) {
        LOG.info("report fail url {}", url);
        boolean force = false;
        for (UrlStats urlStats : urlList) {
            if (url.startsWith(urlStats.url)) {
                if (urlStats.active.compareAndSet(true, false)) {
                    force = true;
                }
            }
        }
        probe(force);
    }

    public void probe(boolean force) {
        if (!force && (System.currentTimeMillis() - lastProbeTime.get() < probe_min_period_ms)) {
            return;
        }
        if (!isRunning.compareAndSet(false, true)) {
            LOG.info("test is already running");
            return;
        }
        Runnable runThread = () -> test();
        runThread.run();
    }

    private List<String> getUrlList(String url) {
        try {
            List<String> newUrlList = probeFunction.getUrlList(url);
            if (newUrlList != null && !newUrlList.isEmpty()) {
                return newUrlList;
            }
        } catch (Exception e) {
            LOG.error("exception to get URL list from {}", url, e);
        }
        return null;
    }

    private List<String> refreshUrlList() {
        List<String> newUrlList = null;
        for (UrlStats urlStats : urlList) {
            if (urlStats.active.get()) {
                newUrlList = getUrlList(urlStats.url);
                if (newUrlList != null) {
                    return newUrlList;
                }
            }
        }
        // try from the inactive URLs
        for (UrlStats urlStats : urlList) {
            newUrlList = getUrlList(urlStats.url);
            if (newUrlList != null) {
                return newUrlList;
            }
        }
        return newUrlList;
    }

    private void test() {
        try {
            List<String> newUrlList = refreshUrlList();
            if (newUrlList == null || newUrlList.isEmpty()) {
                LOG.error("failed to get new URL list!");
                return;
            }
            if (newUrlList.size() == 1) {
                return;
            }
            List<UrlStats> results = new ArrayList<>();
            for (String url : newUrlList) {
                long startTime = System.currentTimeMillis();
                try {
                    List<String> tmpList = probeFunction.getUrlList(url);
                    if (tmpList != null && !tmpList.isEmpty()) {
                        LOG.info("got one active url {}", url);
                        results.add(new UrlStats(url, System.currentTimeMillis() - startTime, true));
                    } else {
                        LOG.error("failed to get url list from url {}", url);
                    }
                } catch (Exception e) {
                    LOG.error("exception to probe url {}", url, e);
                }
            }
            if (!results.isEmpty()) {
                results.sort(Comparator.comparingLong(UrlStats::getResponseTime));
                urlList = results;
            }
        } catch (Exception e) {
            LOG.error("exception to probe URLs!", e);
        } finally {
            isRunning.set(false);
            lastProbeTime.set(System.currentTimeMillis());
        }
    }
}

File: sdk/pc-sdk-go/utils/url_probe.go
/*
 * Copyright (c) 2025 Yangagile. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package utils

import (
	"log"
	"sort"
	"sync"
	"sync/atomic"
	"time"
)

// URL Probe
type UrlProbe struct {
	probeMinPeriodMs int64
	urlListPtr       atomic.Value //  *[]*UrlStats
	probeFunction    UrlProbeFunction
	isRunning        int32
	lastProbeTime    int64
	baseUrl          string
}

type UrlProbeFunction func(url string) ([]string, error)

type UrlStats struct {
	url          string
	responseTime int64
	active       int32
}

func NewUrlProbe(baseUrl string, probeFunction UrlProbeFunction) (*UrlProbe, error) {
	urls, err := probeFunction(baseUrl)
	if err != nil {
		log.Printf("Failed to get URL list from %s: %v", baseUrl, err)
		urls = []string{baseUrl}
		return nil, err
	}

	urlList := make([]*UrlStats, 0, len(urls))
	for _, url := range urls {
		urlList = append(urlList, &UrlStats{
			url:          url,
			responseTime: 0,
			active:       1,
		})
	}
	probe := UrlProbe{
		probeMinPeriodMs: 5 * 60 * 1000,
		probeFunction:    probeFunction,
		isRunning:        0,
		lastProbeTime:    0,
		baseUrl:          baseUrl,
	}
	probe.urlListPtr.Store(&urlList)
	return &probe, nil
}

func (u *UrlProbe) getUrlList() []*UrlStats {
	ptr := u.urlListPtr.Load()
	if ptr == nil {
		return nil
	}
	return *(ptr.(*[]*UrlStats))
}

func (u *UrlProbe) setUrlList(urlList []*UrlStats) {
	// 创建新的切片指针
	newPtr := &urlList
	u.urlListPtr.Store(newPtr)
}

// SetProbeMinPeriodMs
func (u *UrlProbe) SetProbeMinPeriodMs(period int64) {
	atomic.StoreInt64(&u.probeMinPeriodMs, period)
}

func (u *UrlProbe) GetUrl() string {
	urlList := u.getUrlList()
	if urlList == nil || len(urlList) == 0 {
		u.probe(true)
		return ""
	}

	// 首先尝试活跃的URL
	for _, urlStats := range urlList {
		if atomic.LoadInt32(&urlStats.active) == 1 {
			return urlStats.url
		}
	}

	// 如果没有活跃的URL，返回第一个
	if len(urlList) > 0 {
		return urlList[0].url
	}

	u.probe(false)
	return ""
}

// ReportFail 报告URL失败
func (u *UrlProbe) ReportFail(url string) {
	log.Printf("Report fail url: %s", url)
	force := false

	currentList := u.getUrlList()
	if currentList == nil {
		return
	}

	for _, urlStats := range currentList {
		if len(url) >= len(urlStats.url) && url[:len(urlStats.url)] == urlStats.url {
			if atomic.CompareAndSwapInt32(&urlStats.active, 1, 0) {
				force = true
			}
		}
	}

	if force {
		u.probe(true)
	}
}

// probe 触发探测
func (u *UrlProbe) probe(force bool) {
	currentTime := time.Now().UnixMilli()
	lastProbeTime := atomic.LoadInt64(&u.lastProbeTime)

	if !force && (currentTime-lastProbeTime < atomic.LoadInt64(&u.probeMinPeriodMs)) {
		return
	}

	if !atomic.CompareAndSwapInt32(&u.isRunning, 0, 1) {
		log.Println("Test is already running")
		return
	}

	// probe async
	go u.test()
}

// refresh UrlList
func (u *UrlProbe) refreshUrlList() []string {
	currentList := u.getUrlList()

	if currentList != nil {
		// try active urls
		for _, urlStats := range currentList {
			if atomic.LoadInt32(&urlStats.active) == 1 {
				urls, err := u.probeFunction(urlStats.url)
				if err == nil && len(urls) > 0 {
					return urls
				}
			}
		}

		// try all urls
		for _, urlStats := range currentList {
			urls, err := u.probeFunction(urlStats.url)
			if err == nil && len(urls) > 0 {
				return urls
			}
		}
	}

	// try base url
	urls, err := u.probeFunction(u.baseUrl)
	if err == nil && len(urls) > 0 {
		return urls
	}

	return []string{}
}

// probe the URLs and update the active status
func (u *UrlProbe) test() {
	defer func() {
		atomic.StoreInt32(&u.isRunning, 0)
		atomic.StoreInt64(&u.lastProbeTime, time.Now().UnixMilli())
	}()

	newUrlList := u.refreshUrlList()
	if len(newUrlList) == 0 {
		log.Println("Failed to get new URL list!")
		return
	}

	if len(newUrlList) == 1 {
		return
	}

	results := make([]*UrlStats, 0, len(newUrlList))
	var wg sync.WaitGroup
	var resultMutex sync.Mutex

	for _, url := range newUrlList {
		wg.Add(1)
		go func(targetUrl string) {
			defer wg.Done()

			startTime := time.Now()
			_, err := u.probeFunction(targetUrl)
			responseTime := time.Since(startTime).Milliseconds()

			resultMutex.Lock()
			defer resultMutex.Unlock()

			if err == nil {
				log.Printf("Got one active url: %s", targetUrl)
				results = append(results, &UrlStats{
					url:          targetUrl,
					responseTime: responseTime,
					active:       1,
				})
			} else {
				log.Printf("Failed to probe url %s: %v", targetUrl, err)
			}
		}(url)
	}

	wg.Wait()

	if len(results) > 0 {
		// sort by response time
		sort.Slice(results, func(i, j int) bool {
			return results[i].responseTime < results[j].responseTime
		})
		u.setUrlList(results)
	}
}

// GetActiveUrls 获取所有活跃的URL（用于调试和监控）
func (u *UrlProbe) GetActiveUrls() []string {
	currentList := u.getUrlList()

	activeUrls := make([]string, 0)
	for _, urlStats := range currentList {
		if atomic.LoadInt32(&urlStats.active) == 1 {
			activeUrls = append(activeUrls, urlStats.url)
		}
	}
	return activeUrls
}
